### ISSUE TITLE
No error when using await with arbitrary types

### ISSUE THREAD

**Issue by marcin-chwedczuk** [Microsoft/TypeScript#26880]
Today I spot this problem while doing a code review. I was really taken aback that TS compiler emitted no errors.  

Awaiting on arbitrary value like `!step2()` in this example (that results in `true`/`false` value) should produce an error since it doesn't makes any sense and also produces wrong programs. Also notice that all functions in my example are typed so this problem should be easy to fix (since all type info is already there).

```ts
// runnable in TS Playground

async function step1(): Promise<void> {
    console.log('before - step1')
    var x = await !step2(); // WRONG HERE
    console.log('after x= ' + x + ' - step1')
}

async function step2(): Promise<boolean> {
    console.log('before - step2')
    var result = await someApiCall();
    console.log('after - step2 (result = ' + result + ')')
    return (result == 'meow');
}  

function someApiCall(): Promise<string> {
    // simulate api call
    return new Promise((resolve, reject) => {
        setTimeout(
            () => resolve('meow'),
            1000);
    });
}

step1()
    .catch(err => console.log('ERROR TOP LEVEL: ' + err.stack));
```

**Comment by ajafff** [comment]
This is working as intended. The specification allow `await`ing any value. If the awaited value implements the Promise protocol, it awaits the result of that Promise. Otherwise it resolves a new Promise with the value and awaits that promise.

----

If you want to avoid `await` of non-Promise values, you can use a linter to detect such cases.

TSLint has a rule `await-promise` that *tries* to do what you want. But be aware it has [certain bugs or limitations](https://github.com/palantir/tslint/issues/3997#issuecomment-400767267).

I implemented a similar rule for my own linter runtime [Fimbullinter](https://github.com/fimbullinter/wotan). It's called [await-only-promise](https://github.com/fimbullinter/wotan/blob/master/packages/mimir/docs/await-only-promise.md) and is basically the better version of TSLint's `await-promise`. [Click here](https://github.com/fimbullinter/wotan/blob/master/packages/wotan/README.md) for documentation about CLI usage, configuration and more.

**Comment by andy-ms** [comment]
There is a tslint [rule](https://palantir.github.io/tslint/rules/await-promise/) for this.
There's also [strict-boolean-expressions](https://palantir.github.io/tslint/rules/strict-boolean-expressions/) which would warn on `!step2()` whether that's awaited or not.

**Comment by nattthebear** [comment]
> This is working as intended. The specification allow awaiting any value. 

The specification also allows things like `string - number` which typescript rejects as it almost always indicates an error.  Is there an intentional use case for `await unknown`?

**Comment by andy-ms** [comment]
Duplicate of #13376

**Comment by RyanCavanaugh** [comment]
It's certainly reasonable to write `await expr` when `expr` is e.g. `string | Promise<string>`. The usual rule for union types is that an operation must be well-defined for each of its constituents, so the question then becomes one of consistency vs intuitiveness.
